design-proposals/instancetype.kubevirt.io/extend-preferSpread.md

```diff
@@ -0,0 +1,325 @@
+
+For example:
+
+```yaml
+
+$ ./cluster-up/kubectl.sh apply -f -<<EOF
+---
+apiVersion: instancetype.kubevirt.io/v1beta1
+kind: VirtualMachineInstancetype
+metadata:
+  name: spread
+spec:
+  cpu:
+    guest: 4
+  memory:
+    guest: 1Gi
+---
+apiVersion: instancetype.kubevirt.io/v1beta1
+kind: VirtualMachinePreference
+metadata:
+  name: spread
+spec:
+  cpu:
+    preferredCPUTopology: preferSpread
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-spread
+spec:
+  instancetype:
+    name: spread
+    kind: VirtualMachineInstancetype
+  preference:
+    name: spread
+    kind: VirtualMachinePreference
+  runStrategy: Always
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+        - containerDisk:
+            image: quay.io/containerdisks/fedora:39
+          name: containerdisk
+EOF
+[..]
+$ ./cluster-up/kubectl.sh get vmi/vm-spread -o json | jq .spec.domain.cpu
+{
+  "cores": 2,
+  "model": "host-model",
+  "sockets": 2,
+  "threads": 1
+}
+```
+
+This design proposal aims to extend this implementation by introducing new configurables within a preference to control how vCPUs are spread with the goal of allowing vCPUs to also be spread to threads.
+
+## Motivation
+
+The current implementation of the `preferSpread` `preferredCPUTopology` is limited to spreading vCPUs across sockets and cores of the guest CPU topology. While a useful starting point it would be beneficial to also allow these vCPUs to be spread across all parts of the guest visible CPU topology including threads in order to mimic SMT within the guest.
+
+This is particularly useful for performance related workloads such as DPDK etc where we would like to make use of SMT on the host in order to map through sibling threads from pCPUs to guest thread siblings. Note that this use case requires `dedicatedCPUPlacement` and `guestMappingPassthrough` to be useful so the CPUManager can allocate complete pCPUs (via the `full-pcpus-only` option) and for these pCPU siblings to be mapped correctly as vCPU thread siblings to the guest.
+
+## Goals
+
+* Allow for vCPUs to be spread across all attributes of the guest visible CPU topology
+
+## User Stories
+
+* As an admin or VM owner I would like to provide preferences that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+* As a VM owner I would like to consume preferences that that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+## Repos
+
+* kubevirt/kubevirt
+
+* kubevirt/common-instancetypes
+
+# Design
+
+### `spec.cpu.spreadOptions`
+
+A new `spreadOptions` struct will be introduced to hold new configurables detailing how vCPUs should be spread over the guest visible CPU topology.
```

Comment:
In which API? I suspect in the VMI?

Reply:
Apologies this is within the preference spec, I'll make that clear.